src/encdatavault_fmt_plug.c

```diff
@@ -0,0 +1,434 @@
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted.
+ */
+
+#if FMT_EXTERNS_H
+extern struct fmt_main fmt_encdadatavault;
+#elif FMT_REGISTERS_H
+john_register_one(&fmt_encdadatavault);
+#else
+
+#include <string.h>
+
+#ifdef _OPENMP
+#include <omp.h>
+#endif
+
+#include "aes.h"
+#include "md5.h"
+#include "common.h"
+#include "formats.h"
+
+#define FORMAT_TAG              "$encdv$"
+#define FORMAT_TAG_LENGTH       (sizeof(FORMAT_TAG) - 1)
+#define FORMAT_LABEL            "ENCDataVault"
+#define FORMAT_NAME             ""
+#define ALGORITHM_NAME          "MD5"
+#define BENCHMARK_COMMENT       " (1000 iterations)"
+#define BENCHMARK_LENGTH        0x107
+#define PLAINTEXT_LENGTH        125
+#define HEX_BINARY_SIZE         (REAL_BINARY_SIZE*2)
+#define BINARY_SIZE             0
+#define BINARY_ALIGN            1
+#define SALT_SIZE               sizeof(struct custom_salt)
+#define SALT_ALIGN              4
+#define MIN_KEYS_PER_CRYPT      1
+#define MAX_KEYS_PER_CRYPT      4
+
+#ifndef OMP_SCALE
+#define OMP_SCALE               16
+#endif
+
+#define ENC_DEFAULT_NUM_HASH_ITERATIONS 1000
+#define ENC_SALT_SIZE 16
+#define ENC_IV_SIZE 16
+#define ENC_BLOCK_SIZE 16
+#define ENC_KEY_SIZE 16
+#define ENC_NONCE_SIZE 8
+#define ENC_SIG_SIZE 8
+#define ENC_MAX_KEY_NUM 8
+#define ENC_KEYCHAIN_SIZE 128
+
+typedef union buffer_128_u {
+	uint8_t u8[8];
+	uint64_t u64[2];
+} buffer_128;
+
+static MAYBE_INLINE void enc_xor_block(uint64_t *dst, const uint64_t *src)
+{
+	dst[0] ^= src[0];
+	dst[1] ^= src[1];
+}
+
+// md5("f46fcf4f0d9d45b198eb240ff819aac0\x00\x00\x00" + index)
+static const char *default_salts[] = {
+	"\x0f\xc9\xe7\xd0\x8b\xe4\x24\xf6\x56\x9d\x4e\x72\xed\xbc\x2c\x5c",
+	"\xdd\x79\x74\xf3\x3d\x83\x00\xc2\x9b\xd2\x93\xd5\x7f\x9d\x9b\x8c",
+	"\x60\x85\x0c\x47\x58\x46\xe2\x96\x2d\x99\x5d\x5e\xf1\xd0\x6a\x28",
+	"\xe2\x3f\x3d\x6b\x99\x61\x4b\xa9\xc4\xed\xc5\xdd\xd8\x25\x3c\xe1",
+	"\x2c\xa4\x59\x89\x1d\x78\x52\xdb\x30\x31\xd0\x9f\x9f\x34\x88\x35",
+	"\xdb\x1b\xb5\x27\xe8\x21\x4f\x79\xa0\xb2\xcb\x32\x42\xd9\xf2\x0a",
+	"\xae\xa8\xb6\x8e\xd0\x7b\x62\xa1\x40\x0e\x17\xc6\xad\x64\x20\xc8",
+	"\xea\xe3\xf4\x4e\xaf\x4a\x8f\x84\xf1\xfa\xb3\x08\x85\x69\xbe\xf8"
+};
+
+static struct fmt_tests encdatavault_tests[] = {
+	// Sandisk vault
+	{ "$encdv$1$1$7531d98593ea7b1d$d070650d68d50e6e", "openwall" },
+	// Sony vault
+	{ "$encdv$1$2$4e277e5547b88f9c$eb0ed650ee5dffc6", "openwallopenwall!" },
+	// ENCDataVault 512 bits user password
+	{ "$encdv$1$3$d6209d17c0a87818$77608f044a4f113a", "123456789ABCDEf" },
+	// ENCDataVault 1024 bits vault
+	{ "$encdv$3$4$bc93a92cf625e360$fe120bfb658d27b6$739f9d85964cef2f63b927ff77f3328cc5192014ae21c29954c322f4e808fe5c8abe64cc150dbcb08cb334f3b5c28357f10d8d5c6a103e2c899402136c14633aaf8a5347959b33b80ade2e3f5698864605940dc1423704999e5da859d6584491bbe940f00f162c75d2766b868ef2b4c6bf599a2e8ccea3f7cdfab193744a8d10", "123456789ABCDEf" },
```

Comment:
Ideally, we'd have 2+ test vectors per type (so 8+ total). There are some kinds of errors that having only one test vector per type wouldn't catch but 2+ sometimes would.

Reply:
Indeed I spotted a bug while adding tests, thank you !

Reply:
FWIW, you still only have one test vector for each one prefix of `$1$3` and `$3$*`. Maybe adding another to each of those would uncover more bugs. ;-)

Reply:
BTW, there's also `--test-full=1` to run more elaborate tests than the default.

Reply:
I did not know `--test-full=1`, that's nice. I added the tests.